### ISSUE TITLE
Increasing capacity of AutoHashMap should invalidate pointers

### ISSUE THREAD

**Issue by BogdanFloris** [ziglang/zig#18198]
### Zig Version

0.12.0-dev.1717+54f4abae2

### Steps to Reproduce and Observed Behavior

Hello 👋🏻 I've encountered an issue today while using Zig. Not sure if this is a bug or expected behaviour.

Here's a reproducible example:

```
const std = @import("std");

var gpa = std.heap.GeneralPurposeAllocator(.{}){};
const allocator = gpa.allocator();

pub fn main() !void {
    var map = std.AutoHashMap(u64, u64).init(allocator);
    defer map.deinit();
    for (0..1000) |i| {
        const value_at_i = try map.getOrPutValue(i, 1);
        for (i..i + 10) |j| {
            // If this makes an allocation, value_at_i can point to random memory
            // and the += will segfault.
            const value_at_j = try map.getOrPutValue(j, 1);
            value_at_j.value_ptr.* += value_at_i.value_ptr.*;
        }
    }
}
```

TLDR: When `try map.getOrPutValue(j, 1)` does not have enough memory, the map is grown and thus `value_at_i` may point to random memory. When we try to make the increment on the next line, this will segfault.

### Expected Behavior

I've gotten around this by redeclaring the pointer after the array was grown in the inner for loop, but I think the compiler should detect that there is possible grow in the inner for-loop and not allow me to reuse the `value_at_i` pointer. Let me know what you think! Thank you!

**Comment by BogdanFloris** [comment]
Cool, that makes sense! I didn't think about a feature where the compiler would detect if the address is no longer allocated. Maybe a feature where the compiler can detect if I am using `value_at_i` after the map may have possibly been grown/modified. But I realise that's hard to implement as well.

Good to know it's expected behaviour. Thank you! Closing this.
